x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/SearchHitRowSet.java

```diff
@@ -102,15 +110,61 @@ protected Object extractValue(HitExtractor e) {
         for (int lvl = 0; lvl <= extractorLevel ; lvl++) {
             // TODO: add support for multi-nested doc
             if (hit != null) {
-                SearchHits innerHits = hit.getInnerHits().get(innerHit);
-                sh = innerHits == null ? SearchHits.EMPTY : innerHits.getHits();
+                SearchHit[] innerHits = flatInnerHits.get(hit).get(innerHit);
+                sh = innerHits == null ? SearchHits.EMPTY : innerHits;
             }
             hit = sh[indexPerLevel[lvl]];
         }
 
         return e.extract(hit);
     }
 
+    private SearchHit[] getAllInnerHits(SearchHit hit, String path) {
+        if (hit == null) {
+            return null;
+        }
+        List<SearchHits> list = hit.getInnerHits().entrySet().stream().filter(entry -> {
+            try {
+                // the name of an inner_hits section is "[path]_[number]", so keep all inner_hits matching this rule
+                Long.valueOf(entry.getKey().substring(path.length() + 1));
+            } catch(NumberFormatException nfe) {
```

Comment:
What's the case that this happens?
And if there is, can't it result in `StringIndexOutOfBoundsException` because of the `path.length() + 1` ?

Reply:
I think it is possible for a scenario where there are two nested fields: one has the path `pathA` and the other has the path `pathA_`. Looking for elements matching `pathA_[number]` means `startsWith(pathA)` is `true` and what's left after `pathA_` should be a number. In the case of `pathA__34` (second path inner_hits name) - the first condition (startsWith) is true, but the second condition is not and the `pathA__34` is filtered out. Can you think of another way in which I can check the pattern `path_[number]`?

Reply:
I think you could simply use:
```
int endOfPath = s.lastIndexOf('_');
return entry.getKey().substring(0, endOfPath).equals(path)
```
and remove the parsing of the trailing number